### PR TITLE
Make name resolution errors non-fatal

### PR DESCRIPTION
**Pull request by nrc**
We can now handle name resolution errors and get past type checking (if we're a bit lucky). This is the first step towards doing code completion for partial programs (we need error recovery in the parser and early access to save-analysis).

### CODE REVIEW

**Comment by nrc** [comment]
r? @nikomatsakis 

cc @rust-lang/compiler anyone can review this, I don't think it needs any special knowledge. I'd be grateful if anyone has the time.

File: src/test/compile-fail/resolve-inconsistent-binding-mode.rs [hunk 1]
```diff
@@ -16,6 +16,7 @@ fn matcher1(x: opts) {
     match x {
       opts::a(ref i) | opts::b(i) => {}
       //~^ ERROR variable `i` is bound with different mode in pattern #2 than in pattern #1
+      //~^^ ERROR mismatched types
```
Comment (eddyb) [inline src/test/compile-fail/resolve-inconsistent-binding-mode.rs]: Sounds like these should be kept as typeck errors and removed from resolve.

**Comment by nikomatsakis** [comment]
This looks good to me. I agree with @eddyb that reporting an error in the case of `ref i` vs `i` is unfortunate -- it feels like a "downstream error" which would be best to suppress, though it might be a pain. Could we move that "mode" checking into typeck easily enough?

**Comment by nikomatsakis** [comment]
r+ modulo that question.

**Comment by nrc** [comment]
Seems easy enough to change to only giving a type mismatch error in that case (I hacked the change and we didn't fail any tests). However the error message is a fair bit less informative and it seems tricky to add that information to the type mismatch error, so in this case I'd be in favour of leaving the duplicate errors (or trying to suppress the type mismatch, but that is trickier and best left as a follow-up).

**Comment by nikomatsakis** [comment]
@nrc 

> Seems easy enough to change to only giving a type mismatch error in that case (I hacked the change and we didn't fail any tests). However the error message is a fair bit less informative and it seems tricky to add that information to the type mismatch error, so in this case I'd be in favour of leaving the duplicate errors (or trying to suppress the type mismatch, but that is trickier and best left as a follow-up).

I was thinking of trying to suppress the type mismatch, yes. OK.

**Pull request merged**